### PR TITLE
(Reload Users 3/3) return unnissued users in Ajax callback and inject them

### PR DESCRIPTION
**Pull request by octaviorequena**
Source: [NTGR-673](https://accredible.atlassian.net/browse/NTGR-673)

### CODE REVIEW

File: locallib.php [hunk 1]
```diff
@@ -22,6 +22,9 @@
  * @copyright  Accredible <dev@accredible.com>
  * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
+defined('MOODLE_INTERNAL') || die();
+
+require_once($CFG->dirroot . '/mod/quiz/locallib.php');
```
Comment (octaviorequena) [inline locallib.php:27]: This is needed so the method `quiz_get_best_grade` doesn't fail when calling this method from the external function.

File: locallib.php [hunk 2]
```diff
@@ -31,28 +34,19 @@
...
-function accredible_check_if_cert_earned($record, $course, $user) {
+function accredible_check_if_cert_earned($record, $user) {
     global $DB;
 
     $earned = false;
 
     // Check for the existence of an activity instance and an auto-issue rule.
     if ( $record and ($record->finalquiz or $record->completionactivities) ) {
 
-        // Check if we have a groupid or achievementid. Logic is same for both.
```
Comment (octaviorequena) [inline locallib.php:46]: Remove as `$groupid` is not used anywhere.

File: mod_form.php [hunk 3]
```diff
@@ -182,32 +185,32 @@ public function definition() {
...
 
         // Manually issue certificates header.
         $mform->addElement('header', 'chooseusers', get_string('manualheader', 'accredible'));
         // Hidden message to be displayed with Javascript when no users are available.
-        $mform->addElement('html', '<div id="users-warning" class="hidden">');
+        $mform->addElement('html', '<div class="manual-issue-warning hidden">');
         $mform->addElement('static', 'nouserswarning', '', get_string('nouserswarning', 'accredible'));
         $mform->addElement('html', '</div>');
-        $mform->addElement('html', '<div id="users-container">');
         $this->add_checkbox_controller(1, 'Select All/None');
```
Comment (kei178) [inline mod_form.php:212]: Minor - This is just a small thing and can be ignored, but the state of `Select All/None` won't be reset when the selected group is changed. It means if the user clicks `Select All/None` to select all users and change the group, the user will have to click `Select All/None` twice to select all users again.

Comment (kei178) [inline mod_form.php:212]: Minor - This is just a small thing and can be ignored but the state of `Select All/None` won't be reset when the selected group is changed. It means if the user clicks `Select All/None` to select all users and change the group, the user will have to click `Select All/None` twice to select all users again.

**Review by kei178 — COMMENTED** [review]
@octaviorequena - I left several comments.

**Review by Ndibui — APPROVED** [review]
`js` looks good.

File: tests/mod_accredible_locallib_test.php [hunk 4]
```diff
@@ -145,6 +145,74 @@ public function test_accredible_get_transcript() {
...
+
+        // When user has attempted and pass the quiz.
+        $quiz = $this->create_quiz_module(1, $this->user->id);
+        $this->create_quiz_grades($quiz->id, $this->user->id, 10);
+        $accredible = $this->create_accredible_instance($this->course->id, $quiz->id);
+        $result = accredible_check_if_cert_earned($accredible, $user);
+
+        // It responds with false.
+        $this->assertEquals($result, true);
+    }
```
[the remark this answers is not in the document]
Comment (octaviorequena) [reply tests/mod_accredible_locallib_test.php:191]: As explained to you before, we need to review this logic as I think it's not working correctly.

[the remark this answers is not in the document]
Comment (kei178) [reply tests/mod_accredible_locallib_test.php:191]: The logic is not working so we decided to leave it as it is for now. We'll consider how to fix it in [NTGR-704](https://accredible.atlassian.net/browse/NTGR-704).

File: amd/src/userlist_updater.js [hunk 5]
```diff
@@ -29,86 +29,128 @@ define(['jquery', 'core/ajax', 'core/templates'], function($, Ajax, Templates) {
...
                 t.userwarning.addClass('hidden');
-                t.userscontainer.removeClass('hidden');
+                t.manualuserscontainer.removeClass('hidden');
+                t.unissuedusersmessage.removeClass('hidden');
+                t.selectallbutton.removeClass('hidden');
             }
         },
 
+        /**
+         * Hide users and display warning when not group is selected.
```
Comment (naveenbalachandar) [inline amd/src/userlist_updater.js:75]: Typo - "Hide users and display warning when no group is selected."

File: amd/src/userlist_updater.js [hunk 6]
```diff
@@ -29,86 +29,128 @@ define(['jquery', 'core/ajax', 'core/templates'], function($, Ajax, Templates) {
...
+         */
+        displayNotUsersWarning: function() {
```
Comment (naveenbalachandar) [inline amd/src/userlist_updater.js:77]: I think this should be `displayNoUsersWarning`?

**Review by naveenbalachandar — APPROVED** [review]

**Pull request merged**